chall08/gbudau.py

```diff
@@ -0,0 +1,173 @@
+       not one_exit(minefield):
+        return False
+    return True
+
+def check_inner_grid(minefield):
+    robots = 0
+
+    for row in minefield[1:-1]:
+        robots += row[1:-1].count(_ROBOT)
+        if not all(c in _INNER_GRID for c in row[1:-1]):
+            return False
+    return robots == 1
+
+def validate_minefield(minefield):
+    if not minefield or len(minefield) < 3 or \
+       len(minefield[0]) < 3 or not equal_width(minefield) or \
+       not check_outer_grid(minefield) or not check_inner_grid(minefield):
+        return False
+    return True
+
+def validate_actions(actions):
+    if not actions or not all(c in _VALID_ACTIONS for c in actions):
+        return False
+    return True
+
+def robot_position(minefield):
+    for x, row in enumerate(minefield):
+        for y, col in enumerate(row):
+            if col in _ROBOT:
+                return x, y
+    return -1, -1
+
+def move_robot_exit(minefield, action, row, col):
+    height_idx = len(minefield) - 1
+    width_idx = len(minefield[0]) - 1
+
+    if action in "N" and row > 0 and col in range(1, width_idx):
+        row -= 1
+    elif action in "S" and row < height_idx and col in range(1, width_idx):
+        row += 1
+    elif action in "W" and col > 0 and row in range(1, height_idx):
+        col -= 1
+    elif action in "E" and col < width_idx and  row in range(1, height_idx):
+        col += 1
+    return row, col
+
+def move_robot(minefield, action, row, col):
+    if action in "N" and minefield[row - 1][col] not in _WALL:
+        row -= 1
+    elif action in "S" and minefield[row + 1][col] not in _WALL:
+        row += 1
+    elif action in "W" and minefield[row][col - 1] not in _WALL:
+        col -= 1
+    elif action in "E" and minefield[row][col + 1] not in _WALL:
+        col += 1
+    return row, col
+
+def remote_control(minefield, actions):
+    on = False
+
+    row, col = robot_position(minefield)
+    if row == -1 or col == -1:
+        return
+    for action in actions:
+        if minefield[row][col] in _MINE:
+            break
+        if action in _INIT and on == False:
+            on = True
+        elif action in _STOP and on == True:
+            on = False
+        elif on == True and minefield[row][col] in _EXIT:
+            row, col = move_robot_exit(minefield, action, row, col)
+        elif on == True:
+            row, col = move_robot(minefield, action, row, col)
+
+    if minefield[row][col] in _MINE:
+        print("Exploded.")
+    elif minefield[row][col] in _EXIT and on == False:
+        print("Success.")
+    else:
+        print("Failure.")
+
+def main():
+    try:
+        while True:
+            minefield, err = get_minefield()
+            if err:
+                return
+            actions, err = get_actions()
+            if err:
+                return
+            if not validate_minefield(minefield) or not validate_actions(actions):
+                print("{}: Invalid input.".format(__file__))
```

Comment:
You should exit on `Invalid input.`. You print ``Invalid input.` two times for this test case:

```
+++++++++++
+*0*0*0*0*+
+0*0*0*0*0+
+000000000+
+0000M0000+
+000000000+
++++++++++E
```

Reply:
If this is the test_case7.txt:

```
+++++++++++
+*0*0*0*0*+
+0*0*0*0*0+
+000000000+
+0000M0000+
+000000000+
++++++++++E

ISEEEEEK

asdadasdas
s% 
```

My script is looking for map+newline+instructions then repeating the process until the user press Ctrl-C or Ctrl-D.

For example:

```
+++
+ME
+++

IEK
+++
+ME
+++

IW
```
Will output first 'Success.' then 'Failure.'.

So it works the way i interpreted  the challenge and i misinterpreted the challenge :smile: